### PR TITLE
Use ByteBuf in decrypt job

### PR DESCRIPTION
**Pull request by justinlin-linkedin**
Before this PR, in GetBlobOperation, we use java ByteBuffer as the byte container to store decrypted blob content.
This PR switches to netty ByteBuf.
It also switches to use ioBuffer (direct memory) for encryption and decryption.

### CODE REVIEW

File: ambry-router/src/main/java/com/github/ambry/router/GCMCryptoService.java [hunk 1]
```diff
@@ -123,33 +125,32 @@ public ByteBuf encrypt(ByteBuf toEncrypt, SecretKeySpec key, byte[] iv) throws G
       int outputSize = encrypter.getOutputSize(toEncrypt.readableBytes());
 
       // stick with heap memory for now so to compare with the java.nio.ByteBuffer.
```
Comment (zzmao) [inline ambry-router/src/main/java/com/github/ambry/router/GCMCryptoService.java:127]: No longer valid.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 2]
```diff
@@ -1353,12 +1356,19 @@ private void handleMetadataBlob(BlobData blobData, byte[] userMetadata, ByteBuff
...
+                      result.getDecryptedBlobContent().release();
+                    }
+                  } else {
+                    decryptJobMetricsTracker.onJobCallbackProcessingStart();
+                    logger.trace("Handling decrypt job call back for Metadata chunk {} to set decrypt callback results",
+                        blobId);
+                    decryptCallbackResultInfo.setResultAndException(result, exception);
+                    routerCallback.onPollReady();
+                    decryptJobMetricsTracker.onJobCallbackProcessingComplete();
+                    processCallbacks();
```
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1370]: i'm a little worried about calling processCallbacks from a different thread in a callback. I don't think it was designed to be called from non event loop threads. Will take a deeper look to see if there's any alternative (or if it is already thread safe)

Comment (justinlin-linkedin) [reply ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1370]: you are right, the progresstracker and someother object is not thread safe and should probably only be used in event loop thread. I will update it.

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 3]
```diff
@@ -1354,12 +1357,19 @@ private void handleMetadataBlob(BlobData blobData, byte[] userMetadata, ByteBuff
...
```
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1371]: Apologies if you already briefed me on this, but what was the original reason for moving the process callback logic to the callback? Does this prevent a specific leak situation?

File: ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java [hunk 4]
```diff
@@ -1139,49 +1142,49 @@ MessageFormatFlags getOperationFlag() {
...
+            decryptCallbackResultInfo.exception);
+        setOperationException(
+            new RouterException("Exception thrown on decrypting content for " + blobType + " blob " + blobId,
+                decryptCallbackResultInfo.exception, RouterErrorCode.UnexpectedInternalError));
+        progressTracker.setCryptoJobFailed();
+      } else {
+        // in case of Metadata blob, only user-metadata needs decryption if the blob is encrypted
+        if (blobType == BlobType.MetadataBlob) {
+          logger.trace("Processing stored decryption callback result for Metadata blob {}", blobId);
+          initializeDataChunks();
```
Comment (cgtz) [inline ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java:1159]: For 100% read-after-write across thread guarantee, the variables set in `initializeDataChunks()`, `blobInfo`, etc. should probably be volatile (possible I am misinterpreting the documentation). However, I suspect that we have a few other cases where this was not done which have not caused (discoverable) issues in practice :P 

https://docs.oracle.com/javase/specs/jls/se8/html/jls-17.html#jls-17.4.4

**Comment by justinlin-linkedin** [comment]
Hi @cgtz  I revert the thread change for decryption callback, for I found another way to avoid a corner case bytebuf leak. Now the decryption callback is in the same thread as poll and handleResponse.

**Review by cgtz — APPROVED** [review]

**Pull request merged**